Added Animation, Tilemaps, and many more (#24)
twolivesleft/Codea4-Docs: docs/source/api/tilemap.rst

Reviewer: Lots of helptext missing here as well.

tilemap
=======

> Api for creating tile maps

Reviewer: ```suggestion
API for creating tile maps.
```

> **It Contains**

Reviewer: See my animation.rst comment about this.

>     - ``tm.tiles`` - infomation on a single tile

Reviewer: ```suggestion
    - ``tm.tiles`` - information on a single tile
```

>     - ``tm.ruleset`` - infomation on how sprites should behavior in a tile

Reviewer: ```suggestion
    - ``tm.ruleset`` - infomartion on how sprites should behave in a tile
```

    - ``tm.tileset`` - collection of tiles used in the scene
    - ``tm.layer`` - the placement of tiles in the single layer
    - ``tm.tilemap`` - the grouping of tilemap layers to draw to the scene

.. lua:module:: tm

Tile
####

.. lua:class:: tile

    .. code-block:: lua
        :caption: How to use tile

        myTileset = tm.tileset()

        newTile = myTileset:tile()

        newTile:sprite(asset.dirtTile):group(7):collision(tm.collision.square)
            :ruleset(myRuleSet)

        spriteImg = newTile:sprite()
        theRuleset = newTile:ruleset()

    .. lua:attribute:: id: number

>     .. lua:method:: sprite([spriteIcon])

Reviewer: Look at other cases in the Codea4-Docs. I believe we have another way to document these optional parameter functions.

        :param spriteIcon: The image that the tile contains
        :type spriteIcon: sprite
      
        No parameter:

        :return: The sprite image
        :rtype: sprite

    .. lua:method:: group(groupNum)

        Set the sprite image of the tile

        :param groupNum: The group number the tile is from
        :type groupNum: number

        :return: self for function chaining
        :rtype: tile

    .. lua:method:: collision(mode)

        Set the collision mode of the tile

        :param mode: The enum of the collision
        :type mode: enum

        :return: self for function chaining
        :rtype: tile

        **Collision Mode Enum:**

        * ``tm.collision.none`` - no collision
        * ``tm.collision.square`` - for square collision
        * ``tm.collision.sprite`` - for sprite collision

    .. lua:method:: ruleset([theRuleset])

        Set/Get the ruleset of the tile

        :param theRuleset: The ruleset to be applied to the tile
        :type theRuleset: tm.ruleset
      
        No parameter:

        :return: The ruleset of the tile
        :rtype: tm.ruleset

Ruleset
#######

.. lua:class:: ruleset

>     A ruleset is a object to allows the user to determine how the same tiles should a aranged using certain rules.

Reviewer: ```suggestion
    A ruleset is an object to determine how the same tiles should arrange using certain rules.
```

>     Having a ruleset simplify the creation of tilemap as common patterns can be set as a rule

Reviewer: I don't understand this sentence.

    .. lua:method:: rule()

>         Creates a rule in the ruleset and select it. Following ruleset functions will apply to this rule.

Reviewer: ```suggestion
        Creates a rule in the ruleset and selects it. Following ruleset functions will apply to this rule.
```

>         Creates a rule in the ruleset and select it. Following ruleset functions will apply to this rule.

Reviewer: As a general comment, when you say "selects it", I think you are referring to the chaining? If so, it's not really selecting anything, it's just returning it, so I think using "returns" instead of "selects" in all cases would make more sense.

        :return: self for function chaining. 
        :rtype: ruleset

    .. lua:method:: [index] (ruleNum)

        Select the rule in the ruleset. Following ruleset functions will apply to this rule.

        :param ruleNum: the index of the rule in the ruleset
        :type ruleNum: integer

        :return: self for function chaining
        :rtype: ruleset

    **Below happens to rule created above**

    .. lua:method:: sprite([spriteIcon])

>         Set/Get the sprite image of the rule

Reviewer: I think we generally use Get/Set and not Set/Get?

        :param spriteIcon: The image that the rule contains
        :type spriteIcon: sprite
      
        No parameter:

        :return: One sprite image for regular and a table for `random`
        :rtype: sprite or table<sprite>

    .. lua:method:: random([spriteList])

>         Set the sprite that will be randomly selected (good for dirt tiles)

Reviewer: Not sure it's worth mentioning dirt tiles here. It's good for pretty much any type of world tile...

        :param spriteList: The images that the rule contains
        :type spriteList: table<sprite>
      
        :return: self for function chaining
        :rtype: ruleset

    .. lua:method:: area(row1,... , rowX)

>         Set the tile area rule to determine with sprite should be display in the correct spot. The rows must be a old number 3 - 7. Rows and cols should be the same length

Reviewer: ```suggestion
        Set the tile area rule to determine with sprite should be displayed in the correct spot. The rows must be a old number 3 - 7. Rows and cols should be the same length
```

Reviewer: What do you mean here by old? odd? and if so, why "3 - 7"? That looks like a range.
